File: 2025/10/python/main.py
import numpy as np
from scipy.optimize import linprog

with open("input.txt") as f:
    data = f.readlines()

machines = []

for line in data:
    segments = line.strip().split(" ")
    buttons = segments[1:len(segments) - 1]
    buttons = [b.removeprefix("(").removesuffix(")").split(",") for b in buttons]
    buttons = [[int(d) for d in b] for b in buttons]

    joltage = segments[-1].removeprefix("{").removesuffix("}").split(",")
    joltage = [int(j) for j in joltage]

    machines.append({"buttons": buttons, "joltage": joltage})

total = 0

for machine in machines:
    objective_coeffs = np.ones(len(machine["buttons"]))

    columns = []
    for button in machine["buttons"]:
        initial = [0] * len(machine["joltage"])
        for index in button:
            initial[index] = 1
        columns.append(initial)

    A_eq = np.array(columns).T

    b_eq = np.array(machine["joltage"])

    bounds = [(0, float("inf"))] * len(machine["buttons"])

    result = linprog(c=objective_coeffs,
                     A_eq=A_eq,
                     b_eq=b_eq,
                     bounds=bounds,
                     integrality=[1] * len(machine["buttons"])
                     )
    total += int(sum(result.x))

print("Puzzle 2 result:", int(total))

File: 2025/10/kotlin/src/Main.kt
import java.nio.file.Files
import java.nio.file.Path

data class Machine(
    val lights: List<Boolean>,
    val buttons: List<List<Int>>,
    val joltageRequirements: List<Int>,
)

fun puzzle1(machines: List<Machine>) {
    val puzzle1 = machines.sumOf { solveMachineIndicatorLights(it) }
    println("Puzzle 1 result: $puzzle1")
}

fun solveMachineIndicatorLights(machine: Machine): Int {
    val expected = machine.lights
    val visited = mutableMapOf(expected.map { false } to 0)
    val queue = mutableListOf(expected.map { false })
    while (true) {
        val current = queue.removeFirst()
        val depth = visited[current]!!
        machine.buttons.forEach { button ->
            val nextDepth = depth + 1
            val nextState = current.mapIndexed { index, light ->
                if (button.contains(index)) {
                    !light
                } else {
                    light
                }
            }
            if (nextState == expected) {
                return nextDepth
            }
            if (visited.contains(nextState)) {
                return@forEach
            }
            visited[nextState] = nextDepth
            queue.add(nextState)
        }
    }
}

fun main() {
    val file = Files.readString(Path.of("input.txt"))
    val machines = file.lines()
        .map { line ->
            val parts = line.split(" ")
            val lights = parts.first().removePrefix("[").removeSuffix("]").map {
                when (it) {
                    '.' -> false
                    '#' -> true
                    else -> throw IllegalArgumentException("Unsupported light state $it")
                }
            }
            val buttons = parts.subList(1, parts.lastIndex).map { button ->
                button.removePrefix("(").removeSuffix(")").split(",").map { it.toInt() }
            }
            val joltageRequirements = parts.last().removePrefix("{").removeSuffix("}").split(",").map { it.toInt() }
            Machine(
                lights,
                buttons,
                joltageRequirements
            )
        }
    puzzle1(machines)
}
